receiver/k8sclusterreceiver/receiver.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package k8sclusterreceiver
+
+import (
+	"context"
+	"time"
+
+	"github.com/open-telemetry/opentelemetry-collector/component"
+	"github.com/open-telemetry/opentelemetry-collector/consumer"
+	"go.uber.org/zap"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+)
+
+var _ component.MetricsReceiver = (*kubernetesReceiver)(nil)
+
+type kubernetesReceiver struct {
+	resourceWatcher *resourceWatcher
+
+	config   *Config
+	logger   *zap.Logger
+	consumer consumer.MetricsConsumerOld
+	cancel   context.CancelFunc
+}
+
+func (kr *kubernetesReceiver) Start(ctx context.Context, host component.Host) error {
+	var c context.Context
+	c, kr.cancel = context.WithCancel(ctx)
+
+	go func() {
+		kr.resourceWatcher.startWatchingResources(c.Done())
+
+		ticker := time.NewTicker(kr.config.CollectionInterval)
+		defer ticker.Stop()
+
+		for {
+			select {
+			case <-ticker.C:
+				kr.dispatchMetricData(ctx)
+			case <-c.Done():
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (kr *kubernetesReceiver) Shutdown(context.Context) error {
+	kr.cancel()
+	return nil
+}
+
+func (kr *kubernetesReceiver) dispatchMetricData(ctx context.Context) {
+	for _, m := range kr.resourceWatcher.dataCollector.CollectMetricData() {
+		kr.consumer.ConsumeMetricsData(ctx, m)
```

Comment:
This can return an error. It needs to be handled using `obsreport`. See for example [prometheus](https://github.com/open-telemetry/opentelemetry-collector/blob/048d7d2578249a4cd1ad083568688fc389792ea4/receiver/prometheusreceiver/internal/transaction.go#L143) receiver or [opencensus](https://github.com/open-telemetry/opentelemetry-collector/blob/048d7d2578249a4cd1ad083568688fc389792ea4/receiver/opencensusreceiver/ocmetrics/opencensus.go#L129) receiver.
If we don't do this it will be silently ignored.

Reply:
Updated to handle errors returned by ConsumeMetricsData. I noticed the other receivers look like the handling of these errors are done per `ConsumeMetricsData` call but in those cases, it also seems like the metrics are from the same source? For this receiver metrics originate from various sources, so I've clubbed errors from all `ConsumeMetricsData` calls in an interval. Let me know what you think.

Also, these internal metrics about number of timeseries and number of datapoints, are these send to the exporter too?